docs/Tutorials/Configuration.md
## Configuration

The default configuration file is shown as follows.

```sh
{
  "database": {
    "mysql": {
      "username": "username",
      "password": "password",
      "database": "mysql",
      "host": "localhost",
      "port": 3306,
      "dialect": "mysql",
      "operatorsAliases": false,
      "logging": false,
      "timezone": "+08:00",
      "define": {
        "freezeTableName": true,
        "timestamp": true,
        "charset": "utf8"
      },
      "pool": {
        "max": 10,
        "min": 1,
        "idle": 10000,
        "acquire": 30000
      }
    },
    "redis": {
      "cluster": false,
      "options": [
        {
          "host": "localhost",
          "port": 6379
        }
      ]
    },
    "mongodb": {
      "host": "localhost",
      "port": 27017,
      "db": "loraLogger",
      "cluster": false
    }
  },
  "mqClient_ns": {
    "consumerGroup": {
      "options": {
        "kafkaHost": "localhost:9092",
        "groupId": "lora-network-server-message-dispatch-in",
        "sessionTimeout": 15000,
        "protocol": [
          "roundrobin"
        ],
        "fromOffset": "latest"
      },
      "topics": [
        "NS-sub",
        "AS-pub",
        "JS-pub"
      ]
    },
    "client": {
      "kafkaHost": "localhost:9092",
      "clientId": "lora-network-server-message-dispatch-out"
    },
    "producer": {
      "requireAcks": 1,
      "ackTimeoutMs": 100,
      "partitionerType": 2
    },
    "schemaPath": {
      "messages": "config/messages.json",
      "common": "config/common.json"
    },
    "topics": {
      "pubToApplicationServer": "AS-sub",
      "subFromApplicationServer": "AS-pub",
      "pubToConnector": "NC-sub",
      "subFromConnector": "NS-sub",
      "pubToJoinServer": "JS-sub",
      "subFromJoinServer": "JS-pub",
      "pubToControllerServer": "CS-sub",
      "subFromControllerServer": "CS-pub"
    }
  },
  "log": {
    "level": "debug",
    "colorize": true
  },
  "server": {
    "fcntCheckEnable": true,
    "deduplication_Delay": 200,
    "downlink_Data_Delay": 200
  },
  "mqClient_js": {
    "consumerGroup": {
      "options": {
        "kafkaHost": "localhost:9092",
        "groupId": "lora-join-server-consumer",
        "sessionTimeout": 15000,
        "protocol": [
          "roundrobin"
        ],
        "fromOffset": "latest"
      },
      "topics": [
        "JS-sub"
      ]
    },
    "client": {
      "kafkaHost": "localhost:9092",
      "clientId": "lora-join-server-produce"
    },
    "producer": {
      "requireAcks": 1,
      "ackTimeoutMs": 100,
      "partitionerType": 2,
      "joinServerTopic": "JS-pub"
    }
  },
  "mqClient_nc": {
    "consumerGroup": {
      "options": {
        "kafkaHost": "localhost:9092",
        "groupId": "lora-network-connector-consumer",
        "sessionTimeout": 15000,
        "protocol": [
          "roundrobin"
        ],
        "fromOffset": "latest"
      },
      "topics": [
        "NC-sub"
      ]
    },
    "client": {
      "kafkaHost": "localhost:9092",
      "clientId": "lora-network-connector-produce"
    },
    "producer": {
      "requireAcks": 1,
      "ackTimeoutMs": 100,
      "partitionerType": 2
    },
    "topics": {
      "pubToServer": "NS-sub"
    }
  },
  "udp": {
    "port": 1700
  },
  "http": {
    "port": 3000
  },
  "mqClient_as": {
    "consumerGroup": {
      "options": {
        "kafkaHost": "localhost:9092",
        "groupId": "lora-application-server-message-dispatch-in",
        "sessionTimeout": 15000,
        "protocol": [
          "roundrobin"
        ],
        "fromOffset": "latest"
      },
      "topics": [
        "AS-sub",
        "cloud-sub"
      ]
    },
    "client": {
      "kafkaHost": "localhost:9092",
      "clientId": "lora-application-server-message-dispatch-out"
    },
    "producer": {
      "requireAcks": 1,
      "ackTimeoutMs": 100,
      "partitionerType": 2
    },
    "schemaPath": {
      "messages": "config/messages.json",
      "common": "config/common.json"
    },
    "topics": {
      "pubToCloud": "cloud-pub",
      "subFromCloud": "cloud-sub",
      "pubToServer": "AS-pub",
      "subFromServer": "AS-sub"
    }
  },
  "mqClient_nct": {
    "consumerGroup": {
      "options": {
        "kafkaHost": "localhost:9092",
        "groupId": "lora-network-controller-message-dispatch-in",
        "sessionTimeout": 15000,
        "protocol": [
          "roundrobin"
        ],
        "fromOffset": "latest"
      },
      "topics": [
        "CS-sub"
      ]
    },
    "client": {
      "kafkaHost": "localhost:9092",
      "clientId": "lora-network-controller-message-dispatch-out"
    },
    "producer": {
      "requireAcks": 1,
      "ackTimeoutMs": 100,
      "partitionerType": 2
    },
    "schemaPath": {
      "messages": "config/messages.json",
      "common": "config/common.json"
    },
    "topics": {
      "pubToServer": "CS-pub",
      "subFromServer": "CS-sub"
    }
  },
}
```

In general, the following options are necessary for users to modify according to actual situations,

- database.mysql.username: the username of MySQL
- database.mysql.password: the user's password of MySQL
- database.mysql.database: the database used for X-LoRa in MySQL
- udp.port: the port for receiving UDP packets from LoRa Gateways(default value is 1700)
- http.port: the port of HTTP interfaces(default value is 3000)

Other configurations are free to change to fit the environments.

---


config.js
'use strict';

//Network-server identifier
const nsid = 1;
const kafkaHost = 'localhost:9092';

module.exports = {

  //Database settings
  database: {
    mysql: {
      username: 'username',
      password: 'password',
      database: 'mysql',
      host: 'localhost',
      port: 3306,
      dialect: 'mysql',
      operatorsAliases: false,
      logging: false,
      timezone: '+08:00',
      define: {
        'freezeTableName': true,
        'timestamp': true,
        'charset': 'utf8',
      },
    },
    redis: {
      cluster: false,
      options: [{
        host: 'localhost',
        port: 6379,
        retryStrategy: function (times) {
          var delay = Math.min(times * 50, 30000);
          if (delay >= 30000) {
            console.log('---------------Redis Retry close---------------');
            return 'close';
          }
          return delay;
        }
      }],
    },
    mongodb: {
      host: 'localhost',
      port: 27017,
      db: 'loraLogger',
      cluster: false,
    },
  },

  //Test file setting
  mocha: {
    timeout: 5000,
    longTimeOut: 15000,
  },

  //Kafka consumer & producer setting
  mqClient_ns: {
    nsid: `${nsid}`, // if exist in topic schema
    consumerGroup: {
      options: {
        kafkaHost: kafkaHost,
        groupId: `lora-network-server-message-dispatch-in-${nsid}`,
        sessionTimeout: 15000,
        protocol: ['roundrobin'],
        fromOffset: 'latest'
      },
      topics: ['NS-sub', 'AS-pub', 'JS-pub']
    },
    client: {
      kafkaHost: kafkaHost,
      clientId: `lora-network-server-message-dispatch-out-${nsid}`
    },
    producer: {
      requireAcks: 1,
      ackTimeoutMs: 100,
      partitionerType: 2
    },

    //Custom shema to verify packet validity. Not used in server.
    schemaPath: {
      messages: 'config/messages.json',
      common: 'config/common.json'
    },

    //Kafka topics for receiving and sending data
    topics: {
      pubToApplicationServer: 'AS-sub',
      subFromApplicationServer: 'AS-pub',
      pubToConnector: 'NC-sub',
      subFromConnector: 'NS-sub',
      pubToJoinServer: 'JS-sub',
      subFromJoinServer: 'JS-pub',
      pubToControllerServer: 'CS-sub',
      subFromControllerServer: 'CS-pub'
    },
  },

  //winston logger setting
  log: {
    level: 'debug',
    colorize: true,
  },

  //LoRa network-server setting
  server: {
    /*
    Set whether to enable FCnt check
    */
    fcntCheckEnable: true,
    /*
    Waiting time for Uplink package de-duplication
    Unit: milliseconds(ms)
    */
    deduplication_Delay: 200,
    /*
    Waiting time for downlink package processing
    Unit: milliseconds(ms)  
    */
    downlink_Data_Delay: 200,//ms
  },

  mqClient_js: {
    consumerGroup: {
      options: {
        kafkaHost: kafkaHost,
        groupId: 'lora-join-server-consumer',
        sessionTimeout: 15000,
        protocol: ['roundrobin'],
        fromOffset: 'latest'
      },
      topics: [
        'JS-sub'
      ]
    },
    client: {
      kafkaHost: kafkaHost,
      clientId: 'lora-join-server-producer'
    },
    producer: {
      requireAcks: 1,
      ackTimeoutMs: 100,
      partitionerType: 2,
      joinServerTopic: 'JS-pub'
    }
  },

  mqClient_nc: {
    consumerGroup: {
      options: {
        kafkaHost: kafkaHost,
        groupId: 'lora-network-connector-consumer',
        sessionTimeout: 15000,
        protocol: ['roundrobin'],
        fromOffset: 'latest'
      },
      topics: [
        'NC-sub'
      ]
    },
    client: {
      kafkaHost: kafkaHost,
      clientId: 'lora-network-connector-producer'
    },
    producer: {
      requireAcks: 1,
      ackTimeoutMs: 100,
      partitionerType: 2
    },
    topics: {
      pubToServer: 'NS-sub'
    }
  },

  mqClient_as: {
    nsid: `${nsid}`, // if exist in topic schema
    consumerGroup: {
      options: {
        kafkaHost: kafkaHost,
        groupId: `lora-application-server-message-dispatch-in-${nsid}`,
        sessionTimeout: 15000,
        protocol: ['roundrobin'],
        fromOffset: 'latest'
      },
      topics: ['AS-sub', `cloud-sub-${nsid}-lora`]
    },
    client: {
      kafkaHost: kafkaHost,
      clientId: `lora-application-server-message-dispatch-out-${nsid}`
    },
    producer: {
      requireAcks: 1,
      ackTimeoutMs: 100,
      partitionerType: 2
    },
    schemaPath: {
      messages: 'config/messages.json',
      common: 'config/common.json'
    },
    topics: {
      pubToCloud: `cloud-pub-${nsid}`,
      subFromCloud: `cloud-sub-${nsid}-lora`,
      pubToServer: 'AS-pub',
      subFromServer: 'AS-sub',
    },
  },

  pb: {
    '260febaf6c60807b': 'pm25_message.pm25',
  }
};
